### DISCUSSION TITLE
when useRouter can't return value immediately, how can I avoid the error message from fetching.

### DISCUSSION THREAD

**Discussion by LuYujin9** [neuefische/questions#571] [Web] [answered]
Hi, could you please help me?
This is my code:
```javascript

  const router = useRouter();
  const { id } = router.query;
  const { isReady } = router;

  const { data: reserve, isLoading, error } = useSWR(`/api/reserves/${id}`);
  const { data: restaurant } = useSWR(
    `/api/restaurants/${reserve?.restaurantId}`
  );
  const { trigger: triggerRestaurant } = useSWRMutation(
    `/api/restaurants/${reserve?.restaurantId}`,
    updateData
  );
  const { trigger: triggerReserve } = useSWRMutation(
    `/api/reserves/${id}`,
    updateData
  );
  if (!isReady || !restaurant || isLoading || error) return <h2>Loading</h2>;
``` 

Because the `router` can't get the value in the first render, and the hooks can't be behind '  if (!isReady || !restaurant || isLoading || error) return `<h2>Loading</h2>;`. Before the value of  `id`  accessed, the hooks return the error:
/  capstone-project-9ouw5d35j-luyujin9.vercel.app/api/restaurants/undefined 500

After the `router` finally gets the value, all the things work.  when people don't see the console, the website looks normal. But I really curious, how can the problem be fixed.

the preview link :
https://capstone-project-9ouw5d35j-luyujin9.vercel.app/my-data/reserves/6438307b44f5e5cffbad99ce

my code:
https://github.com/LuYujin9/capstone-project/blob/edit-the-reserve/pages/my-data/reserves/%5Bid%5D.js

**Comment by PReichetanz** [comment]
Hi @LuYujin9,

I'm not completely sure whether I get your problem right (and I was not able to reproduce it properly), but I've found one issue regarding the `isReady` key of the `router` object in [the official docs](https://nextjs.org/docs/api-reference/next/router#router-object):

> isReady: boolean - Whether the router fields are updated client-side and ready for use. Should only be used inside of useEffect methods and not for conditionally rendering on the server. See related docs for use case with [automatically statically optimized pages](https://nextjs.org/docs/advanced-features/automatic-static-optimization)

I think the point here is the note to use `isReady` only inside of an `useEffect`. I've found [this short blog article](https://thomasstep.com/blog/working-with-nextjs-routers-query) providing a good example how this might look like.

According to these references, I would try to move the `!isReady` from line 31 (`if (!isReady || !restaurant || isLoading || error) return <h2>Loading</h2>;`) into an  ´useEffect` before you need the `id` from `router.query`.

Please let me know if this helps, I'm very curious!

**Reply by LuYujin9** [reply]
Thanks Paul. Anton gave me the answer. I updated my code and don't need to use `isReady` anymore. :)

**Comment by toshydev** [comment, marked as answer]
If I understand it correctly I've had this problem a few times already and my friends from the 🌊 as well. 

I think we solved it with this little addition in the **useSWR** hook:
```const { data } = useSWR(id ? `/api/reserves/${id}` : null)```

Or is this a different issue?

**Reply by LuYujin9** [reply]
Thanks Anton. That is exactly my problem! It works now.
